File: java/src/test/java/myairtable/tests/TestSerializing.java
package myairtable.tests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.IntNode;
import com.fasterxml.jackson.databind.node.NullNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.time.Instant;
import java.util.Map;
import java.util.Set;
import myairtable.AirtableJson;
import myairtable.PrimaryFields;
import myairtable.PrimaryModel;
import org.junit.jupiter.api.Test;

/**
 * J4.6 — Offline (no-network) model serialization parity with Kotlin TestSerializing / Swift
 * TestSerializing / Rust test_serializing. Network-bound cases live in {@link
 * TestSerializingIntegration} below.
 */
class TestSerializing {

  private static ObjectNode fields() {
    return AirtableJson.MAPPER.createObjectNode();
  }

  private static PrimaryModel decodePrimary(ObjectNode fields) throws Exception {
    return AirtableJson.MAPPER.treeToValue(fields, PrimaryModel.class);
  }

  @Test
  void decodePrimaryFieldsKeyedByFieldId() throws Exception {
    ObjectNode json =
        fields()
            .put(PrimaryFields.primaryKeyId, "PK Value")
            .put(PrimaryFields.singleLineTextId, "hello")
            .put(PrimaryFields.numberIntId, 42)
            .put(PrimaryFields.checkboxId, true);
    PrimaryModel model = decodePrimary(json);
    assertEquals("PK Value", model.getPrimaryKey());
    assertEquals("hello", model.getSingleLineText());
    assertEquals(42.0, model.getNumberInt());
    assertEquals(true, model.getCheckbox());
  }

  @Test
  void dateOnlyAndDatetimeFieldValuesBothDecode() throws Exception {
    ObjectNode json =
        fields()
            .put(PrimaryFields.dateId, "2024-01-15")
            .put(PrimaryFields.dateWithTimeId, "2024-01-15T10:30:00.000Z");
    PrimaryModel model = decodePrimary(json);
    assertEquals(Instant.parse("2024-01-15T00:00:00Z"), model.getDate());
    assertEquals(Instant.parse("2024-01-15T10:30:00Z"), model.getDateWithTime());
  }

  @Test
  void missingFieldsAreNull() throws Exception {
    PrimaryModel model = decodePrimary(fields().put(PrimaryFields.primaryKeyId, "only pk"));
    assertEquals("only pk", model.getPrimaryKey());
    assertNull(model.getSingleLineText());
    assertNull(model.getNumberInt());
    assertNull(model.getDate());
  }

  @Test
  void emptyFieldsObjectDecodesCleanly() throws Exception {
    PrimaryModel model = decodePrimary(fields());
    assertNull(model.getPrimaryKey());
    assertTrue(model.isNew());
  }

  @Test
  void encodeProducesFieldsKeyedByFieldId() {
    PrimaryModel model = PrimaryModel.builder().primaryKey("x").numberInt(7.0).build();
    Map<String, JsonNode> record = model.toCreateFields();
    assertEquals(Set.of(PrimaryFields.primaryKeyId, PrimaryFields.numberIntId), record.keySet());
    assertEquals("x", record.get(PrimaryFields.primaryKeyId).asText());
  }

  @Test
  void roundTripPreservesAllWritableFieldValues() throws Exception {
    PrimaryModel original =
        PrimaryModel.builder()
            .primaryKey("rt")
            .singleLineText("text")
            .checkbox(true)
            .numberInt(5.0)
            .numberFloat(2.5)
            .build();
    String encoded = AirtableJson.MAPPER.writeValueAsString(original);
    PrimaryModel decoded = AirtableJson.MAPPER.readValue(encoded, PrimaryModel.class);
    assertEquals(original.getPrimaryKey(), decoded.getPrimaryKey());
    assertEquals(original.getSingleLineText(), decoded.getSingleLineText());
    assertEquals(original.getCheckbox(), decoded.getCheckbox());
    assertEquals(original.getNumberInt(), decoded.getNumberInt());
    assertEquals(original.getNumberFloat(), decoded.getNumberFloat());
  }

  @Test
  void toRecordEmitsFieldValuesKeyedByFieldId() {
    PrimaryModel model = PrimaryModel.builder().primaryKey("rec").email("a@b.c").build();
    Map<String, JsonNode> record = model.toRecord();
    assertEquals("rec", record.get(PrimaryFields.primaryKeyId).asText());
    assertEquals("a@b.c", record.get(PrimaryFields.emailId).asText());
  }

  @Test
  void freshDecodeHasNoDirtyFieldsAfterSnapshot() throws Exception {
    PrimaryModel model = decodePrimary(fields().put(PrimaryFields.primaryKeyId, "clean"));
    model.takeSnapshot();
    assertTrue(model.dirtyFields().isEmpty());
  }

  @Test
  void mutatingAWritableFieldMarksOnlyThatFieldDirty() throws Exception {
    PrimaryModel model =
        decodePrimary(
            fields()
                .put(PrimaryFields.primaryKeyId, "pk")
                .put(PrimaryFields.singleLineTextId, "before"));
    model.takeSnapshot();
    model.setSingleLineText("after");
    assertEquals(Set.of(PrimaryFields.singleLineTextId), model.dirtyFields().keySet());
  }

  @Test
  void takeSnapshotClearsTheDirtySet() throws Exception {
    PrimaryModel model = decodePrimary(fields().put(PrimaryFields.primaryKeyId, "pk"));
    model.takeSnapshot();
    model.setSingleLineText("mutated");
    assertFalse(model.dirtyFields().isEmpty());
    model.takeSnapshot();
    assertTrue(model.dirtyFields().isEmpty());
  }

  @Test
  void clearedWritableFieldBecomesJsonNull() throws Exception {
    PrimaryModel model =
        decodePrimary(
            fields().put(PrimaryFields.primaryKeyId, "pk").put(PrimaryFields.numberIntId, 9.5));
    model.takeSnapshot();
    model.setNumberInt(null);
    assertEquals(NullNode.getInstance(), model.dirtyFields().get(PrimaryFields.numberIntId));
  }

  @Test
  void unsavedModelReportsIsNewTrue() {
    PrimaryModel model = PrimaryModel.builder().primaryKey("new").build();
    assertTrue(model.isNew());
    assertNull(model.getId());
  }

  @Test
  void modelWithIdReportsIsNewFalse() {
    PrimaryModel model = PrimaryModel.builder().primaryKey("saved").build();
    model.setId("recXYZ");
    assertFalse(model.isNew());
  }

  @Test
  void freshModelOmitsNullFieldsSparseWrite() throws Exception {
    PrimaryModel model = PrimaryModel.builder().primaryKey("sparse").build();
    Map<String, JsonNode> record = model.toCreateFields();
    assertEquals(Set.of(PrimaryFields.primaryKeyId), record.keySet());
    // Jackson-encode also skips nulls (NON_NULL inclusion).
    String encoded = AirtableJson.MAPPER.writeValueAsString(model);
    assertEquals("{\"" + PrimaryFields.primaryKeyId + "\":\"sparse\"}", encoded);
  }

  @Test
  void ratingIsTypeErasedJsonNode() {
    // `rating` has no Airtable metadata type mapping → JsonNode (UNKNOWN).
    PrimaryModel model = PrimaryModel.builder().primaryKey("r").rating(IntNode.valueOf(3)).build();
    assertEquals(IntNode.valueOf(3), model.getRating());
    assertEquals(IntNode.valueOf(3), model.toCreateFields().get(PrimaryFields.ratingId));
  }
}

/**
 * J10.1 — Network-bound serialization cases (split from the offline suite above, parity with Kotlin
 * TestSerializingIntegration / the integration portion of Swift TestSerializing).
 */
class TestSerializingIntegration {
  private final myairtable.Airtable airtable = TestSetup.makeAirtable();

  @Test
  void roundTripPreservesIdAndCreatedTimeAfterFetch() {
    String primaryKey = TestSetup.primaryKey("Serializing", "IdTime");
    PrimaryModel created =
        airtable.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String recordId = created.getId();
    try {
      assertNotNull(created.getCreatedTime(), "createdTime populated on create");
      PrimaryModel fetched = airtable.primary().get(recordId);
      assertEquals(recordId, fetched.getId());
      assertEquals(created.getCreatedTime(), fetched.getCreatedTime());
      airtable.primary().delete(recordId);
    } catch (Throwable e) {
      try {
        airtable.primary().delete(recordId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void linkedRecordFieldsSerializeAsRecordIdArrays() {
    String suite = TestSetup.primaryKey("Serializing", "Links");
    myairtable.SecondaryModel sec =
        airtable
            .secondary()
            .create(myairtable.SecondaryModel.builder().name(suite + " Target").build());
    String secId = sec.getId();
    PrimaryModel prim =
        airtable
            .primary()
            .create(
                PrimaryModel.builder()
                    .primaryKey(suite)
                    .linkSingle(java.util.List.of(secId))
                    .build());
    String primId = prim.getId();
    try {
      Map<String, JsonNode> record = prim.toRecord();
      JsonNode linkElement = record.get(PrimaryFields.linkSingleId);
      assertNotNull(linkElement);
      assertTrue(linkElement.isArray(), "links serialize as arrays");
      assertEquals(secId, linkElement.get(0).asText());
      airtable.primary().delete(primId);
      airtable.secondary().delete(secId);
    } catch (Throwable e) {
      try {
        airtable.primary().delete(primId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      try {
        airtable.secondary().delete(secId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void explicitNullOnAWritableFieldClearsItServerSide() {
    String primaryKey = TestSetup.primaryKey("Serializing", "Clear");
    PrimaryModel created =
        airtable
            .primary()
            .create(
                PrimaryModel.builder()
                    .primaryKey(primaryKey)
                    .singleLineText("to be cleared")
                    .build());
    String recordId = created.getId();
    try {
      created.setSingleLineText(null);
      PrimaryModel saved = airtable.primary().update(created);
      assertNull(saved.getSingleLineText(), "null dirty entry clears the field");

      PrimaryModel fetched = airtable.primary().get(recordId);
      assertNull(fetched.getSingleLineText());
      airtable.primary().delete(recordId);
    } catch (Throwable e) {
      try {
        airtable.primary().delete(recordId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }
}

File: kotlin/src/test/kotlin/myairtable/tests/TestSerializing.kt
package myairtable.tests

import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import myairtable.AirtableJson
import myairtable.PrimaryFields
import myairtable.PrimaryModel
import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * K4.6 — Offline (no-network) model serialization parity with Swift
 * TestSerializing / Rust test_serializing. Network-bound cases (createdTime
 * round-trip, linked records, server-side clears) live in K-F10.
 */
class TestSerializing {
    private val json = AirtableJson.instance

    private fun decodePrimary(builder: kotlinx.serialization.json.JsonObjectBuilder.() -> Unit): PrimaryModel =
        json.decodeFromJsonElement(PrimaryModel.serializer(), buildJsonObject(builder))

    @Test
    fun decodePrimaryFieldsKeyedByFieldId() {
        val model =
            decodePrimary {
                put(PrimaryFields.primaryKeyId, "PK Value")
                put(PrimaryFields.singleLineTextId, "hello")
                put(PrimaryFields.numberIntId, 42)
                put(PrimaryFields.checkboxId, true)
            }
        assertEquals("PK Value", model.primaryKey)
        assertEquals("hello", model.singleLineText)
        assertEquals(42.0, model.numberInt)
        assertEquals(true, model.checkbox)
    }

    @Test
    fun dateOnlyAndDatetimeFieldValuesBothDecode() {
        val model =
            decodePrimary {
                put(PrimaryFields.dateId, "2024-01-15")
                put(PrimaryFields.dateWithTimeId, "2024-01-15T10:30:00.000Z")
            }
        assertEquals(Instant.parse("2024-01-15T00:00:00Z"), model.date)
        assertEquals(Instant.parse("2024-01-15T10:30:00Z"), model.dateWithTime)
    }

    @Test
    fun missingFieldsAreNull() {
        val model = decodePrimary { put(PrimaryFields.primaryKeyId, "only pk") }
        assertEquals("only pk", model.primaryKey)
        assertNull(model.singleLineText)
        assertNull(model.numberInt)
        assertNull(model.date)
    }

    @Test
    fun emptyFieldsObjectDecodesCleanly() {
        val model = decodePrimary { }
        assertNull(model.primaryKey)
        assertTrue(model.isNew)
    }

    @Test
    fun encodeProducesFieldsKeyedByFieldId() {
        val model = PrimaryModel(primaryKey = "x", numberInt = 7.0)
        val record = model.toCreateFields()
        assertEquals(setOf(PrimaryFields.primaryKeyId, PrimaryFields.numberIntId), record.keys)
        assertEquals("x", record[PrimaryFields.primaryKeyId]?.jsonPrimitive?.content)
    }

    @Test
    fun roundTripPreservesAllWritableFieldValues() {
        val original =
            PrimaryModel(
                primaryKey = "rt",
                singleLineText = "text",
                checkbox = true,
                numberInt = 5.0,
                numberFloat = 2.5,
            )
        val encoded = json.encodeToString(PrimaryModel.serializer(), original)
        val decoded = json.decodeFromString(PrimaryModel.serializer(), encoded)
        assertEquals(original.primaryKey, decoded.primaryKey)
        assertEquals(original.singleLineText, decoded.singleLineText)
        assertEquals(original.checkbox, decoded.checkbox)
        assertEquals(original.numberInt, decoded.numberInt)
        assertEquals(original.numberFloat, decoded.numberFloat)
    }

    @Test
    fun toRecordEmitsFieldValuesKeyedByFieldId() {
        val model = PrimaryModel(primaryKey = "rec", email = "a@b.c")
        val record = model.toRecord()
        assertEquals("rec", record[PrimaryFields.primaryKeyId]?.jsonPrimitive?.content)
        assertEquals("a@b.c", record[PrimaryFields.emailId]?.jsonPrimitive?.content)
    }

    @Test
    fun freshDecodeHasNoDirtyFieldsAfterSnapshot() {
        val model = decodePrimary { put(PrimaryFields.primaryKeyId, "clean") }
        model.takeSnapshot()
        assertTrue(model.dirtyFields().isEmpty())
    }

    @Test
    fun mutatingAWritableFieldMarksOnlyThatFieldDirty() {
        val model =
            decodePrimary {
                put(PrimaryFields.primaryKeyId, "pk")
                put(PrimaryFields.singleLineTextId, "before")
            }
        model.takeSnapshot()
        model.singleLineText = "after"
        assertEquals(setOf(PrimaryFields.singleLineTextId), model.dirtyFields().keys)
    }

    @Test
    fun takeSnapshotClearsTheDirtySet() {
        val model = decodePrimary { put(PrimaryFields.primaryKeyId, "pk") }
        model.takeSnapshot()
        model.singleLineText = "mutated"
        assertFalse(model.dirtyFields().isEmpty())
        model.takeSnapshot()
        assertTrue(model.dirtyFields().isEmpty())
    }

    @Test
    fun clearedWritableFieldBecomesJsonNull() {
        val model =
            decodePrimary {
                put(PrimaryFields.primaryKeyId, "pk")
                put(PrimaryFields.numberIntId, 9.5)
            }
        model.takeSnapshot()
        model.numberInt = null
        assertEquals(JsonNull, model.dirtyFields()[PrimaryFields.numberIntId])
    }

    @Test
    fun unsavedModelReportsIsNewTrue() {
        val model = PrimaryModel(primaryKey = "new")
        assertTrue(model.isNew)
        assertNull(model.id)
    }

    @Test
    fun modelWithIdReportsIsNewFalse() {
        val model = PrimaryModel(primaryKey = "saved")
        model.id = "recXYZ"
        assertFalse(model.isNew)
    }

    @Test
    fun freshModelOmitsNullFieldsSparseWrite() {
        val model = PrimaryModel(primaryKey = "sparse")
        val record = model.toCreateFields()
        assertEquals(setOf(PrimaryFields.primaryKeyId), record.keys)
        // Plugin-encode also skips nulls (explicitNulls=false).
        val encoded = json.encodeToString(PrimaryModel.serializer(), model)
        assertEquals("""{"${PrimaryFields.primaryKeyId}":"sparse"}""", encoded)
    }

    @Test
    fun ratingIsTypeErasedJsonElement() {
        // `rating` has no Airtable metadata type mapping → JsonElement (UNKNOWN).
        val model = PrimaryModel(primaryKey = "r", rating = JsonPrimitive(3))
        assertEquals(JsonPrimitive(3), model.rating)
        assertEquals(JsonPrimitive(3), model.toCreateFields()[PrimaryFields.ratingId])
    }
}
